crates/build/src/solidity_metadata/abi.rs

```diff
@@ -0,0 +1,391 @@
+///
+/// Ref: <https://docs.soliditylang.org/en/latest/abi-spec.html#abi-json>
+pub fn write_abi<P>(abi: &JsonAbi, path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let json = serde_json::to_string(abi)?;
+    fs::write(path, json)?;
+
+    Ok(())
+}
+
+/// Returns the constructor ABI representation for an ink! constructor.
+fn constructor(
+    ctor: &ConstructorSpec<PortableForm>,
+    registry: &PortableRegistry,
+) -> Result<Constructor> {
+    let params = ctor
+        .args()
+        .iter()
+        .map(|param| {
+            param_decl(param, registry, &format!("constructor `{}`", ctor.label()))
+        })
+        .process_results(|mut iter| iter.join(","))?;
+
+    // NOTE: Solidity constructors don't expose a return type.
+    let abi_str = format!(
+        "constructor({params}){}",
+        if ctor.payable { " payable" } else { "" }
+    );
+    Constructor::parse(&abi_str).map_err(|err| {
+        anyhow::anyhow!(
+            "Failed to parse abi for constructor `{}` : {err}",
+            ctor.label()
+        )
+    })
+}
+
+/// Returns the function ABI representation for an ink! message.
+fn message(
+    msg: &MessageSpec<PortableForm>,
+    registry: &PortableRegistry,
+) -> Result<Function> {
+    let name = msg.label();
+    let params = msg
+        .args()
+        .iter()
+        .map(|param| param_decl(param, registry, &format!("message `{}`", name)))
+        .process_results(|mut iter| iter.join(","))?;
+    let ret_ty = return_ty(msg.return_type(), registry, &format!("message `{}`", name))?;
+
+    let abi_str = format!(
+        "function {name}({params}) public{}{}{}",
+        // FIXME: (@davidsemakula) ink! does NOT currently enforce it's immutability
+        // claims for messages intrinsically (i.e at compile time).
+        // Ref: <https://github.com/use-ink/ink/issues/1969>
+        if msg.mutates() { "" } else { " view" },
+        if msg.payable() { " payable" } else { "" },
+        if ret_ty.is_empty() || ret_ty == "()" {
+            String::new()
+        } else {
+            format!(" returns ({ret_ty})")
+        },
+    );
+    Function::parse(&abi_str).map_err(|err| {
+        anyhow::anyhow!("Failed to parse abi for message `{}` : {err}", msg.label())
+    })
+}
+
+/// Returns the event ABI representation for an ink! event.
+fn event(
+    event_spec: &EventSpec<PortableForm>,
+    registry: &PortableRegistry,
+) -> Result<Event> {
+    let name = event_spec.label();
+    let params = event_spec
+        .args()
+        .iter()
+        .map(|param| {
+            let param_name = param.label();
+            let ty_id = param.ty().ty().id;
+            let sol_ty = resolve_ty(
+                ty_id,
+                registry,
+                &format!("arg `{param_name}` for event `{name}`"),
+            );
+            // TODO: (@davidsemakula) should we simply omit events with Solidity ABI
+            // incompatible types instead of bailing?
```

Comment:
Does this make sense?

Reply:
I would bail, as not being able to persist all events in the metadata breaks the user expectations. If need comes up we could add a flag to overwrite this behavior.

Reply:
Cool ... this is the current behavior.